### PR TITLE
[Ready to Merge] Enable key vault as secret source

### PR DESCRIPTION
**Pull request by Yuqing-cat**
- add key vault dependency
- create a credential client class to init key vault secret client
- add redis sample in credential util which can replace current _construct_redis_client()
- will add more functions and tests if above change is acceptable.

### CODE REVIEW

**Comment by hangfei** [comment]
Please follow https://github.com/linkedin/feathr/blob/main/docs/dev_guide/pull_request_guideline.md.

**Comment by Yuqing-cat** [comment]
What changes to make and why you are making these changes:
- Add akv_client to support reading secrets from Azure Key Vault.
- Key vault logic will be triggered only when environment variable is not set and KEY_VAULT__NAME is given. (envutil)
- Update ARM Template to store REDIS_PASSWORD rather than CONN for better akv usage 

2. How are you going to achieve your goal Describe what testings you have done, for example, performance testing etc.
- So far it will perform exactly the same as it was unless and secret variables are removed and "KEY_VAULT__NAME" is set.

**Comment by xiaoyongzhu** [comment]
@Yuqing-cat let's talk a bit about the scope of this work (probably need a bit bigger scope to cover the spark runtime as well)

File: feathr_project/feathr/_envvariableutil.py [hunk 1]
```diff
@@ -59,5 +61,8 @@ def get_environment_variable(variable_key):
         password = os.environ.get(variable_key)
         if not password:
             logger.info(variable_key +
-                        ' is not set in the environment variables.')
+                        ' is not set in the environment variables. Try Key Vault value.')
```
Comment (jainr) [inline feathr_project/feathr/_envvariableutil.py:64]: When an environment variable is not found, is the expectation for them to set the environment variable with keyVaultName so it is already set?
If it's latter, I suggest modifying the log message to ".....' is not set in the environment variables, fetching the value from Key Vault')

File: feathr_project/feathr/_envvariableutil.py [hunk 2]
```diff
@@ -59,5 +61,8 @@ def get_environment_variable(variable_key):
...
+            akv_name = os.env.get("KEY_VAULT__NAME")
```
Comment (jainr) [inline feathr_project/feathr/_envvariableutil.py:65]: [nitpick] Suggest changing it to KEY_VAULT_NAME , same number of underscores.

File: feathr_project/feathr/akv_client.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from azure.keyvault.secrets import SecretClient
+from azure.identity import DefaultAzureCredential
+from loguru import logger
+from azure.core.exceptions import ResourceNotFoundError
+
+class akvClient:
```
Comment (jainr) [inline feathr_project/feathr/akv_client.py:6]: [nitpick] - Suggest changing it to more descriptive name AzureKeyVaultClient, the way we have for PurviewClient etc.

**Comment by jainr** [comment]
Is KV only storing Redis  Credetials for Online store?  Can we store other credentials too in KV that will unify the implementation when other environment variables are not set.

File: feathr_project/feathr/_envvariableutil.py (hunk 2 above)
Comment (Yuqing-cat) [reply feathr_project/feathr/_envvariableutil.py:65]: was intended to fit the naming pattern of yaml config attributes. But agree, we can use KEY_VAULT_NAME for current functions

File: feathr_project/feathr/_envvariableutil.py (hunk 1 above)
Comment (Yuqing-cat) [reply feathr_project/feathr/_envvariableutil.py]: updated as recommended.

**Review by xiaoyongzhu — APPROVED** [review]

**Review by windoze — APPROVED** [review]

**Pull request merged**